app/models/piece.rb

```diff
@@ -30,20 +41,40 @@ class Piece < ApplicationRecord
 
 
   def vertical_obstruction?(destination_y)
-    range = self.y_position < destination_y ? (self.y_position..destination_y) : (destination_y..self.y_position)
+    range = self.y_position < destination_y ? (self.y_position...destination_y) : (destination_y...self.y_position)
     game.pieces.where(
-        color: self.color,
-        x_position: self.x_position, 
+        # color: self.color,
+        x_position: self.x_position,
         y_position: range
       ).where.not(id: self.id).any?
   end
-  
+
   def horizontal_obstruction?(destination_x)
-    range = self.x_position < destination_x ? (self.x_position..destination_x) : (destination_x..self.x_position)
+    range = self.x_position < destination_x ? (self.x_position...destination_x) : (destination_x...self.x_position)
     game.pieces.where(
-        color: self.color,
-        x_position: range, 
+        # color: self.color,
+        x_position: range,
         y_position: self.y_position
       ).where.not(id: self.id).any?
   end
-end
\ No newline at end of file
+
+  def diagonal_obstruction?(destination_x, destination_y)
+    x_range = self.x_position < destination_x ? (self.x_position...destination_x) : (destination_x...self.x_position)
```

Comment:
(self.x_position...destination_x) - also the destination excluded when it shouldn't be

Reply:
we can't step on the destination square if something is already there